src/network/core/tcp.cpp

```diff
@@ -31,15 +31,19 @@ NetworkTCPSocketHandler::~NetworkTCPSocketHandler()
 {
 	/* Virtual functions get called statically in destructors, so make it explicit to remove any confusion. */
 	this->NetworkTCPSocketHandler::CloseConnection();
+	this->NetworkTCPSocketHandler::CloseSocket();
+}
 
+void NetworkTCPSocketHandler::CloseSocket()
+{
 	if (this->sock != INVALID_SOCKET) closesocket(this->sock);
 	this->sock = INVALID_SOCKET;
 }
 
 NetworkRecvStatus NetworkTCPSocketHandler::CloseConnection(bool error)
 {
+	this->MarkClosed();
 	this->writable = false;
-	NetworkSocketHandler::CloseConnection(error);
 
 	/* Free all pending and partially received packets */
 	while (this->packet_queue != nullptr) {
```

Comment:
Shouldn't the clearing of the packet queue be a separate non-virtual private function? Then `CloseConnection` can just call it, and the destructor too without it being entangled in the virtual-ness of `CloseConnection`. After all, the only thing that is important for the destructor is releasing the packets if there are still any left.